stan/math/opencl/kernel_generator/evaluate_into.hpp

```diff
@@ -0,0 +1,72 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_EVALUATE_INTO_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_EVALUATE_INTO_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_generator/operation.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <cl.hpp>
+#include <utility>
+#include <string>
+#include <set>
+
+namespace stan {
+namespace math {
+
+template <typename Derived, typename ReturnScalar>
+template <typename T_lhs>
+void operation<Derived, ReturnScalar>::evaluate_into(T_lhs&& lhs) const {
+  using enable = enable_if_all_valid_expressions<T_lhs>;
+  using cache = operation<Derived, ReturnScalar>::cache<T_lhs>;
+  auto lhs_expression = as_operation(std::forward<T_lhs>(lhs));
+
+  int n_rows = derived().rows();
+  int n_cols = derived().cols();
+  const char* function = "evaluate_into";
+  if (n_rows != dynamic) {
+    check_size_match(function, "Rows of ", "*this", n_rows, "rows of ",
+                     "lhs_expression", lhs_expression.rows());
+  }
+  if (n_cols != dynamic) {
+    check_size_match(function, "Columns of ", "*this", n_cols, "columns of ",
+                     "lhs_expression", lhs_expression.cols());
+  }
+  try {
+    std::set<int> generated;
+    if (cache::kernel() == NULL) {
+      name_generator ng;
+      kernel_parts parts = derived().generate(generated, ng, "i", "j");
+      kernel_parts out_parts
+          = lhs_expression.generate_lhs(generated, ng, "i", "j");
+      std::string src = "kernel void calculate(" + parts.args +
```

Comment:
That seems like an overkill. Imagine the kernel name for a GLM.

Reply:
Big functions means big signatures! Again the main thing is to imagine debugging stuff. The intent here is that for a particular kernel composition we should be able to trace what is being compounded together. We can get the same thing by keeping meta info about the operations in some base logger

Reply:
I don't think having a huge kernel name would help debugging at all. Intention is that all parts of kernel generator are tested so that if a user makes a bug in his code he does not even have to look at the kernel source. 

If hee wants to check which operations make an expression he is debugging he can check the type of that expression. 

I'm sure if someone really wants to check kernel source he would prefer looking at actual code than some long kernel name.

Also creating a meaningful name would be almost as compliced as generating the kernel source itself. And than we would have to test that it actually matches what the lernel is calculating.

Overall this seems like a lot of work for no gain.